lib/core/base/rule.js

```diff
@@ -228,6 +259,38 @@ Rule.prototype.run = function(context, options, resolve, reject) {
 	q.then(() => resolve(ruleResult)).catch(error => reject(error));
 };
 
+/**
+ * Selects `HTMLElement`s based on configured selector and filters them based on
+ * the rules matches function
+ * @param  {Rule} rule The rule to check for after checks
+ * @param  {Context} context The resolved Context object
+ * @param  {Mixed}   options Options specific to this rule
+ * @return {Array}           All matching `HTMLElement`s
+ */
+function gatherAndMatchNodes(rule, context, options) {
```

Comment:
I don't think this should be a separate function. Feels like a strange architectural decision to pull some logic out of a class method. I get that you're trying to keep this method from bloating, but I'd rather pull line 210 - 239 out into a separate utils function in that case.

I'm not entirely sure about it though, so @stephenmathieson I'd like a second opinion.

Reply:
I agree that moving this logic _out_ of the class doesn't seem like the right decision. Perhaps we just add another method to the class containing this logic and make it more "usable" in all scenarios?

Reply:
When adding the `runSync` method for #1467 , I tried to pull out as much of the duplicate code as possible to make it easier to maintain the two similar functions. This was mostly just an extension of doing that in prep for that pr (as they will conflict)

Reply:
Right. So I think we can make this an instance method that will work for both cases.

Or am I missing something there?

Reply:
I figured since it wasn't callable outside the function of `run`, we wouldn't want to make it now callable publicly. But I can make it a public api method of a rule if we want, it just doesn't do a whole lot on it's own.

Reply:
It sounds like I'm missing something. Let's chat about this tomorrow on Slack.